### DISCUSSION TITLE
[XState v5] How to access event that caused state transition?

### DISCUSSION THREAD

**Discussion by eahenke** [statelyai/xstate#4649] [Q&A] [answered]
I'm working on a router that uses XState as the source of truth, and trying to upgrade it to v5. When a state transition occurs, I need to be able to access the event that caused it in order to build the route and fire the navigation function. In XState v4, I had something like this (simplified):

```
const service = useInterpret(machine);

service.onTransition((state, event) => {
  logTransition(state, event);

  if(IGNORED_EVENTS.has(event.type)) {
    return;
  }
  // State nodes have a meta with a path property like {path: `my-path/:id/` }
  const newPath = getPathFromStateMeta(state);
  // Get event-specific path params like {id: 'my-id'} to build the route
  // Event looks something like: { type: 'EDIT_USER_SCREEN', params: { id: 'my-id' } }
  const pathParams = getPathParams(event);

  // Fire the side effect
  navigate(buildUri(newPath, pathParams));
});
```

[Per the docs](https://stately.ai/docs/migration#use-subscribe-for-changes-not-ontransition), in v5 `onTransition` has been replaced with `subscribe`, but that method does not provide access to the event that caused the transition. I see from #4074 that this is intentional, but the workaround suggested there does not work for my use case (since that was about dev tool logging/debugging and not production behavior).

I also see that there are [transition actors](https://stately.ai/docs/transition-actors) that have access to events, but these are reducers that return state, whereas I am doing a side-effect after the state has changed.

Is there an existing way (or workaround) to fire a side-effect on state change that has access to the event that caused the state change? Or does v5 not support this behavior and I need to change the way this router is designed? Thank you in advance (and apologies if this has already been asked/answered).

**Comment by davidkpiano** [comment, marked as answer]
You can use the [inspect API](https://stately.ai/docs/inspection#snapshot-inspection-events) to listen for events that caused transitions:

```ts
const actor = useActorRef(machine, {
  inspect: (inspEvent) => {
    if (inspEvent.type === '@xstate.snapshot') {
      console.log(inspEvent.event); // the event that caused the transition
      console.log(inspEvent.snapshot); // the resulting snapshot
    }
  }
});
```

**Reply by eahenke** [reply]
Thanks! Somehow I had the impression that this API was just meant for devtools and debugging.

**Reply by chaulkie** [reply]
Hi, I came here with the same question, and the `inspect` API is _okay_, but how can I filter for events that strictly changed the state, instead of the whole snapshot? i.e. I don't care if an event modified the context; only if a state-transition was triggered.
